### PR TITLE
Improve the virtual plugin support

### PR DESCRIPTION
**Pull request by SchSeba**
This commit add the support for virtio interfaces like vhostuser for openstack virtual workers.

implementation details:
* on first run (after a reboot) we get all the information we need when the devices are visible to the kernel
* we match the mac address to the openstack network ID
* if the sriov-network-config-daemon gets reboot it will use the initial file on the node so even if the nics are in vfio the node state will be right
* introduce a new system service to remove the initial state file after a reboot to support nic changes both for virtual and BM environments

Signed-off-by: Sebastian Sch <sebassch@gmail.com>

### CODE REVIEW

**Comment by SchSeba** [comment]
example:

```
apiVersion: v1
items:
- apiVersion: sriovnetwork.openshift.io/v1
  kind: SriovNetworkNodeState
  metadata:
    creationTimestamp: "2022-03-01T15:56:55Z"
    generation: 3
    name: dsal-nfv-mmcqx-worker-0-m8xcm
    namespace: openshift-sriov-network-operator
    ownerReferences:
    - apiVersion: sriovnetwork.openshift.io/v1
      blockOwnerDeletion: true
      controller: true
      kind: SriovNetworkNodePolicy
      name: default
      uid: 4eed7c1f-ae4e-49e6-b165-7955e932ccef
    resourceVersion: "3670266"
    uid: c8c7427c-9e6b-43ec-9da6-d831845e7f52
  spec:
    dpConfigVersion: "3220700"
    interfaces:
    - name: ens4
      numVfs: 1
      pciAddress: "0000:00:04.0"
      vfGroups:
      - deviceType: vfio-pci
        policyName: dpdk1
        resourceName: dpdk1
        vfRange: 0-0
    - numVfs: 1
      pciAddress: "0000:00:06.0"
      vfGroups:
      - deviceType: vfio-pci
        policyName: dpdk2
        resourceName: dpdk2
        vfRange: 0-0
  status:
    interfaces:
    - Vfs:
      - deviceID: "1000"
        driver: virtio-pci
        mac: fa:16:3e:1c:96:9b
        pciAddress: "0000:00:03.0"
        vendor: 1af4
        vfID: 0
      deviceID: "1000"
      driver: virtio-pci
      linkSpeed: -1 Mb/s
      linkType: ETH
      mac: fa:16:3e:1c:96:9b
      name: ens3
      netFilter: openstack/NetworkID:bfe9d7ea-0b50-4e8e-b49f-e051a9a4c1c0
      numVfs: 1
      pciAddress: "0000:00:03.0"
      totalvfs: 1
      vendor: 1af4
    - Vfs:
      - deviceID: "1000"
        driver: vfio-pci
        pciAddress: "0000:00:04.0"
        vendor: 1af4
        vfID: 0
      deviceID: "1000"
      driver: vfio-pci
      netFilter: openstack/NetworkID:32a38a94-9135-497b-9231-afe62a0e3811
      numVfs: 1
      pciAddress: "0000:00:04.0"
      totalvfs: 1
      vendor: 1af4
    - Vfs:
      - deviceID: "1515"
        driver: vfio-pci
        pciAddress: "0000:00:06.0"
        vendor: "8086"
        vfID: 0
      deviceID: "1515"
      driver: vfio-pci
      netFilter: openstack/NetworkID:b4322d87-6032-4247-be7c-9fe00b1b4c97
      numVfs: 1
      pciAddress: "0000:00:06.0"
      totalvfs: 1
      vendor: "8086"
    syncStatus: Succeeded
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

```
apiVersion: v1
items:
- apiVersion: sriovnetwork.openshift.io/v1
  kind: SriovNetworkNodePolicy
  metadata:
    creationTimestamp: "2022-03-01T15:56:52Z"
    generation: 1
    name: default
    namespace: openshift-sriov-network-operator
    resourceVersion: "3218392"
    uid: 4eed7c1f-ae4e-49e6-b165-7955e932ccef
  spec:
    nicSelector: {}
    nodeSelector: {}
    numVfs: 0
    resourceName: ""
- apiVersion: sriovnetwork.openshift.io/v1
  kind: SriovNetworkNodePolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"sriovnetwork.openshift.io/v1","kind":"SriovNetworkNodePolicy","metadata":{"annotations":{},"generation":1,"name":"dpdk1","namespace":"openshift-sriov-network-operator"},"spec":{"deviceType":"vfio-pci","nicSelector":{"netFilter":"openstack/NetworkID:32a38a94-9135-497b-9231-afe62a0e3811"},"nodeSelector":{"feature.node.kubernetes.io/network-sriov.capable":"true"},"numVfs":1,"priority":99,"resourceName":"dpdk1"}}
    creationTimestamp: "2022-03-01T16:00:32Z"
    generation: 1
    name: dpdk1
    namespace: openshift-sriov-network-operator
    resourceVersion: "3220672"
    uid: b71eb562-5118-4353-8530-9d03a7825466
  spec:
    deviceType: vfio-pci
    nicSelector:
      netFilter: openstack/NetworkID:32a38a94-9135-497b-9231-afe62a0e3811
    nodeSelector:
      feature.node.kubernetes.io/network-sriov.capable: "true"
    numVfs: 1
    priority: 99
    resourceName: dpdk1
- apiVersion: sriovnetwork.openshift.io/v1
  kind: SriovNetworkNodePolicy
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"sriovnetwork.openshift.io/v1","kind":"SriovNetworkNodePolicy","metadata":{"annotations":{},"name":"dpdk2","namespace":"openshift-sriov-network-operator"},"spec":{"deviceType":"vfio-pci","nicSelector":{"netFilter":"openstack/NetworkID:b4322d87-6032-4247-be7c-9fe00b1b4c97"},"nodeSelector":{"feature.node.kubernetes.io/network-sriov.capable":"true"},"numVfs":1,"priority":99,"resourceName":"dpdk2"}}
    creationTimestamp: "2022-03-01T16:00:33Z"
    generation: 1
    name: dpdk2
    namespace: openshift-sriov-network-operator
    resourceVersion: "3220681"
    uid: e4e48633-ae01-4247-bbfd-2b06c9a54bf4
  spec:
    deviceType: vfio-pci
    nicSelector:
      netFilter: openstack/NetworkID:b4322d87-6032-4247-be7c-9fe00b1b4c97
    nodeSelector:
      feature.node.kubernetes.io/network-sriov.capable: "true"
    numVfs: 1
    priority: 99
    resourceName: dpdk2
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

File: pkg/daemon/writer.go [hunk 1]
```diff
@@ -200,3 +223,48 @@ func (w *NodeStateStatusWriter) writeCheckpointFile(ns *sriovnetworkv1.SriovNetw
...
+	}
+	defer file.Close()
+	if err = json.NewDecoder(file).Decode(&utils.InitialState); err != nil {
+		return nil, err
+	}
+
+	return &utils.InitialState, nil
+}
+
+func (w *NodeStateStatusWriter) createCleanUpServiceIfNeeded() error {
```
Comment (pliurh) [inline pkg/daemon/writer.go:245]: Do you want this to work on RHCOS nodes? I think we need to use MCO to inject the systemd service for RHCOS nodes.

Comment (SchSeba) [reply pkg/daemon/writer.go:245]: Hi @pliurh I test this one also on RHCOS and it's working the service is created.

another option will be for the sriov operator to create this system service always for masters and workers pools is that a valid solution?

File: pkg/utils/utils_virtual.go [hunk 2]
```diff
@@ -226,6 +294,39 @@ func DiscoverSriovDevicesVirtual(platformType PlatformType, metaData *OSPMetaDat
...
+		names = append(names, f.Name())
+	}
+
+	if len(names) < 1 {
+		return ""
+	}
+
+	return names[0]
+}
+
```
[the remark this answers is not in the document]
Comment (SchSeba) [reply pkg/utils/utils_virtual.go:329]: @bn222 sorry I was not able to understand you will like this function to return (string,error) and just change 

```
if name := tryToGetVirtualInterfaceName(device.Address); name != "" {
```
to 
```
if name,error := tryToGetVirtualInterfaceName(device.Address); err != nil {
```

**Comment by SchSeba** [comment]
rebased on master

**Comment by EmilienM** [comment]
LGTM

**Comment by SchSeba** [comment]
@pliurh can you please have a look?

I change the location of the initial file to be in /tmp on the host file system as requested. this way the file we get removed after every reboot without the need of a system service

File: deploy/configmap.yaml [hunk 3]
```diff
@@ -21,4 +21,4 @@ data:
   Broadcom_bnxt_BCM57414_2x25G: "14e4 16d7 16dc"
   Broadcom_bnxt_BCM75508_2x100G: "14e4 1750 1806"
   Qlogic_qede_QL45000_50G: "1077 1654 1664"
-  
+  Red_Hat_Virtio_network_device: "1af4 1000 1000"
```
[the remark this answers is not in the document]
Comment (MaysaMacedo) [reply deploy/configmap.yaml:24]: can you clarify on the 1000 I see it configured as 0000
https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/271/files#r839536286

**Comment by EmilienM** [comment]
LGTM

File: deploy/configmap.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (SchSeba) [reply deploy/configmap.yaml:24]: I will add it when I rebase the openshift repo base on this PR.

The deviceID is 1000 I checked it again

File: bindata/manifests/config-units/initial-config-reset.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+contents: |
```
Comment (pliurh) [inline bindata/manifests/config-units/initial-config-reset.yaml:1]: We don't need this systemd service anymore, do we?

File: pkg/daemon/writer.go [hunk 5]
```diff
@@ -9,26 +9,27 @@ import (
...
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	snclientset "github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/client/clientset/versioned"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
 )
 
 const (
-	CheckpointFileName = "sno-initial-node-state.json"
+	CheckpointFileName                = "sno-initial-node-state.json"
+	InitialConfigurationResetUnitFile = "bindata/manifests/config-units/initial-config-reset.yaml"
```
Comment (pliurh) [inline pkg/daemon/writer.go:24]: ditto

File: pkg/daemon/writer.go [hunk 6]
```diff
@@ -44,33 +45,49 @@ func NewNodeStateStatusWriter(c snclientset.Interface, n string, f func(), devMo
 
 // Run reads from the writer channel and sets the interface status. It will
 // return if the stop channel is closed. Intended to be run via a goroutine.
-func (writer *NodeStateStatusWriter) Run(stop <-chan struct{}, refresh <-chan Message, syncCh chan<- struct{}, destDir string, runonce bool, platformType utils.PlatformType) {
+func (writer *NodeStateStatusWriter) Run(stop <-chan struct{}, refresh <-chan Message, syncCh chan<- struct{}, destDir string, runonce bool, platformType utils.PlatformType) error {
 	glog.V(0).Infof("Run(): start writer")
 	msg := Message{}
 
-	var err error
+	if runonce {
```
Comment (zeeke) [inline pkg/daemon/writer.go:52]: What about refactoring this if-body to a different function and getting rid of the `runonce` parameters?

**Review by Eoghan1232 — APPROVED** [review]
lgtm

File: bindata/manifests/config-units/initial-config-reset.yaml (hunk 4 above)
Comment (SchSeba) [reply bindata/manifests/config-units/initial-config-reset.yaml]: right! done

File: pkg/daemon/writer.go [hunk 7]
```diff
@@ -42,35 +42,56 @@ func NewNodeStateStatusWriter(c snclientset.Interface, n string, f func(), devMo
...
-		glog.V(0).Info("Run(): once")
-		if err := writer.pollNicStatus(platformType); err != nil {
-			glog.Errorf("Run(): first poll failed: %v", err)
+		if ns == nil {
+			writer.openStackDevicesInfo, err = utils.CreateOpenstackDevicesInfo(metaData, networkData)
+			if err != nil {
+				return err
+			}
+		} else {
+			devicesInfo := make(utils.OSPDevicesInfo)
```
Comment (adrianchiris) [inline pkg/daemon/writer.go:67]: nit: can we move this logic to: `utils.CreateOpenstackDevicesInfoFromNodeStatus(ns)` or similar ?

File: pkg/daemon/writer.go [hunk 8]
```diff
@@ -42,35 +42,56 @@ func NewNodeStateStatusWriter(c snclientset.Interface, n string, f func(), devMo
...
+				devicesInfo[iface.PciAddress] = &utils.OSPDeviceInfo{MacAddress: iface.Mac, NetworkID: iface.NetFilter}
+			}
+			writer.openStackDevicesInfo = devicesInfo
 		}
-		ns, _ := writer.setNodeStateStatus(msg)
-		writer.writeCheckpointFile(ns, destDir)
-		return
 	}
+
+	glog.V(0).Info("RunOnce(): once")
```
Comment (adrianchiris) [inline pkg/daemon/writer.go:75]: is this needed ? there is already a log saying we are in RunOnce at L47

File: pkg/daemon/writer.go [hunk 9]
```diff
@@ -42,35 +42,56 @@ func NewNodeStateStatusWriter(c snclientset.Interface, n string, f func(), devMo
 	}
 }
 
-// Run reads from the writer channel and sets the interface status. It will
-// return if the stop channel is closed. Intended to be run via a goroutine.
-func (writer *NodeStateStatusWriter) Run(stop <-chan struct{}, refresh <-chan Message, syncCh chan<- struct{}, destDir string, runonce bool, platformType utils.PlatformType) {
-	glog.V(0).Infof("Run(): start writer")
+// RunOnce initial the interface status for both baremetal and virtual environments
+func (writer *NodeStateStatusWriter) RunOnce(destDir string, platformType utils.PlatformType) error {
+	glog.V(0).Infof("RunOnce(): start writer")
```
Comment (adrianchiris) [inline pkg/daemon/writer.go:47]: nit: RunOnce(): status writer ? as we are not really starting the writer here

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -459,7 +459,7 @@ func getNetdevMTU(pciAddr string) int {
 }
 
 func getNetDevMac(ifaceName string) string {
-	glog.V(2).Infof("getNetDevMac(): get Mac for device %s", ifaceName)
+	glog.Infof("getNetDevMac(): get Mac for device %s", ifaceName)
```
Comment (adrianchiris) [inline pkg/utils/utils.go:462]: is this logging needed by default (no verbosity) ?

**Review by e0ne — APPROVED** [review]
LGTM

**Review by adrianchiris — COMMENTED** [review]
was accidently logged in as npwg-robot :) LGTM 

I merged this.

**Pull request merged**